--- a/net/minecraft/world/entity/npc/AbstractVillager.java
+++ b/net/minecraft/world/entity/npc/AbstractVillager.java
@@ -164,9 +_,9 @@
    }
 
    @Nullable
-   public Entity m_5489_(ServerLevel p_35295_) {
+   public Entity changeDimension(ServerLevel p_35295_, net.minecraftforge.common.util.ITeleporter teleporter) {
       this.m_7996_();
-      return super.m_5489_(p_35295_);
+      return super.changeDimension(p_35295_, teleporter);
    }
 
    protected void m_7996_() {
@@ -233,5 +_,27 @@
 
    public boolean m_183595_() {
       return this.f_19853_.f_46443_;
+   }
+
+   // Forge Start
+   private net.minecraftforge.common.util.LazyOptional<?> inventoryManager = net.minecraftforge.items.entityinventory.CapabilityEntityInventoryManager.create(this, net.minecraftforge.items.entityinventory.InventoryCarrierInventoryManager::new);
+
+   @Override
+   public <T> net.minecraftforge.common.util.LazyOptional<T> getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable net.minecraft.core.Direction facing) {
+      if (this.m_6084_() && capability == net.minecraftforge.items.entityinventory.CapabilityEntityInventoryManager.ENTITY_INVENTORY_CAPABILITY)
+         return inventoryManager.cast();
+      return super.getCapability(capability, facing);
+   }
+
+   @Override
+   public void invalidateCaps() {
+      super.invalidateCaps();
+      inventoryManager.invalidate();
+   }
+
+   @Override
+   public void reviveCaps() {
+      super.reviveCaps();
+      inventoryManager = net.minecraftforge.items.entityinventory.CapabilityEntityInventoryManager.create(this, net.minecraftforge.items.entityinventory.InventoryCarrierInventoryManager::new);
    }
 }
